File: cpp/tests/test_parents.cpp
#include <catch2/catch_test_macros.hpp>
#include <vost/vost.h>

#include <git2.h>

#include <filesystem>
#include <string>
#include <thread>
#include <chrono>

namespace fs = std::filesystem;

static fs::path make_temp_repo() {
    auto tmp = fs::temp_directory_path() /
               ("vost_ptest_" + std::to_string(
                    std::hash<std::thread::id>{}(std::this_thread::get_id())
                    ^ static_cast<size_t>(
                          std::chrono::steady_clock::now()
                              .time_since_epoch()
                              .count())));
    return tmp;
}

static vost::GitStore open_store(const fs::path& path,
                                  const std::string& branch = "main") {
    vost::OpenOptions opts;
    opts.create = true;
    opts.branch = branch;
    return vost::GitStore::open(path, opts);
}

/// Helper: get the parent count of a commit using libgit2.
static unsigned int parent_count(const fs::path& repo_path,
                                  const std::string& commit_hex) {
    git_repository* repo = nullptr;
    REQUIRE(git_repository_open(&repo, repo_path.c_str()) == 0);

    git_oid oid;
    REQUIRE(git_oid_fromstr(&oid, commit_hex.c_str()) == 0);

    git_commit* commit = nullptr;
    REQUIRE(git_commit_lookup(&commit, repo, &oid) == 0);

    unsigned int count = git_commit_parentcount(commit);
    git_commit_free(commit);
    git_repository_free(repo);
    return count;
}

/// Helper: get the nth parent hash of a commit.
static std::string parent_hash(const fs::path& repo_path,
                                const std::string& commit_hex,
                                unsigned int n) {
    git_repository* repo = nullptr;
    REQUIRE(git_repository_open(&repo, repo_path.c_str()) == 0);

    git_oid oid;
    REQUIRE(git_oid_fromstr(&oid, commit_hex.c_str()) == 0);

    git_commit* commit = nullptr;
    REQUIRE(git_commit_lookup(&commit, repo, &oid) == 0);

    const git_oid* parent_id = git_commit_parent_id(commit, n);
    REQUIRE(parent_id != nullptr);

    char buf[GIT_OID_HEXSZ + 1];
    git_oid_tostr(buf, sizeof(buf), parent_id);
    std::string result(buf, GIT_OID_HEXSZ);

    git_commit_free(commit);
    git_repository_free(repo);
    return result;
}

// ---------------------------------------------------------------------------
// write with parents
// ---------------------------------------------------------------------------

TEST_CASE("Fs: write with advisory parents creates merge commit", "[parents]") {
    auto path  = make_temp_repo();
    auto store = open_store(path);
    auto snap  = store.branches().get("main");

    // Create two branches with different content
    snap = snap.write_text("a.txt", "hello");
    auto branch_a = snap;

    auto snap_b = store.branches().set_and_get("other", snap);
    snap_b = snap_b.write_text("b.txt", "world");

    // Write with advisory parent from branch_b
    vost::WriteOptions opts;
    opts.parents.push_back(*snap_b.commit_hash());
    auto result = branch_a.write_text("merged.txt", "merged", opts);

    // Verify 2 parents
    CHECK(parent_count(path, *result.commit_hash()) == 2);

    // First parent is branch_a tip
    CHECK(parent_hash(path, *result.commit_hash(), 0) == *branch_a.commit_hash());

    // Second parent is snap_b
    CHECK(parent_hash(path, *result.commit_hash(), 1) == *snap_b.commit_hash());

    // Content is correct (no tree merge — just the write)
    CHECK(result.read_text("merged.txt") == "merged");
    CHECK(result.read_text("a.txt") == "hello");

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// batch with parents
// ---------------------------------------------------------------------------

TEST_CASE("Batch: commit with advisory parents", "[parents][batch]") {
    auto path  = make_temp_repo();
    auto store = open_store(path);
    auto snap  = store.branches().get("main");

    snap = snap.write_text("base.txt", "base");

    auto snap_other = store.branches().set_and_get("other", snap);
    snap_other = snap_other.write_text("other.txt", "data");

    vost::BatchOptions bopts;
    bopts.parents.push_back(*snap_other.commit_hash());
    auto batch = snap.batch(bopts);
    batch.write_text("batch.txt", "batch-data");
    auto result = batch.commit();

    CHECK(parent_count(path, *result.commit_hash()) == 2);
    CHECK(parent_hash(path, *result.commit_hash(), 0) == *snap.commit_hash());
    CHECK(parent_hash(path, *result.commit_hash(), 1) == *snap_other.commit_hash());

    CHECK(result.read_text("batch.txt") == "batch-data");
    CHECK(result.read_text("base.txt") == "base");

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// apply with parents
// ---------------------------------------------------------------------------

TEST_CASE("Fs: apply with advisory parents", "[parents]") {
    auto path  = make_temp_repo();
    auto store = open_store(path);
    auto snap  = store.branches().get("main");

    snap = snap.write_text("x.txt", "x");

    auto snap_other = store.branches().set_and_get("feature", snap);
    snap_other = snap_other.write_text("y.txt", "y");

    vost::ApplyOptions aopts;
    aopts.parents.push_back(*snap_other.commit_hash());

    std::vector<std::pair<std::string, vost::WriteEntry>> writes;
    writes.push_back({"z.txt", vost::WriteEntry::from_text("z")});

    auto result = snap.apply(writes, {}, aopts);

    CHECK(parent_count(path, *result.commit_hash()) == 2);
    CHECK(parent_hash(path, *result.commit_hash(), 0) == *snap.commit_hash());
    CHECK(parent_hash(path, *result.commit_hash(), 1) == *snap_other.commit_hash());

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// first-parent lineage preserved
// ---------------------------------------------------------------------------

TEST_CASE("Fs: parent() follows first-parent lineage with advisory parents", "[parents]") {
    auto path  = make_temp_repo();
    auto store = open_store(path);
    auto snap  = store.branches().get("main");

    snap = snap.write_text("a.txt", "a");
    auto first_commit = snap;

    auto snap_other = store.branches().set_and_get("side", snap);
    snap_other = snap_other.write_text("s.txt", "s");

    vost::WriteOptions opts;
    opts.parents.push_back(*snap_other.commit_hash());
    auto merged = first_commit.write_text("b.txt", "b", opts);

    // parent() should return first_commit, not snap_other
    auto p = merged.parent();
    REQUIRE(p.has_value());
    CHECK(p->commit_hash() == first_commit.commit_hash());

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// default: no parents → 1 parent
// ---------------------------------------------------------------------------

TEST_CASE("Fs: write without parents has 1 parent (default)", "[parents]") {
    auto path  = make_temp_repo();
    auto store = open_store(path);
    auto snap  = store.branches().get("main");

    snap = snap.write_text("a.txt", "hello");
    auto result = snap.write_text("b.txt", "world");

    CHECK(parent_count(path, *result.commit_hash()) == 1);
    CHECK(parent_hash(path, *result.commit_hash(), 0) == *snap.commit_hash());

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// multiple advisory parents
// ---------------------------------------------------------------------------

TEST_CASE("Fs: write with multiple advisory parents", "[parents]") {
    auto path  = make_temp_repo();
    auto store = open_store(path);
    auto snap  = store.branches().get("main");

    snap = snap.write_text("a.txt", "a");

    auto b1 = store.branches().set_and_get("b1", snap);
    b1 = b1.write_text("b1.txt", "b1");

    auto b2 = store.branches().set_and_get("b2", snap);
    b2 = b2.write_text("b2.txt", "b2");

    vost::WriteOptions opts;
    opts.parents.push_back(*b1.commit_hash());
    opts.parents.push_back(*b2.commit_hash());
    auto result = snap.write_text("m.txt", "m", opts);

    CHECK(parent_count(path, *result.commit_hash()) == 3);
    CHECK(parent_hash(path, *result.commit_hash(), 0) == *snap.commit_hash());
    CHECK(parent_hash(path, *result.commit_hash(), 1) == *b1.commit_hash());
    CHECK(parent_hash(path, *result.commit_hash(), 2) == *b2.commit_hash());

    fs::remove_all(path);
}

File: rs/tests/test_parents.rs
mod common;

use vost::*;
use vost::fs;

/// Helper: open the bare repo and return the parent count for a commit hash.
fn parent_count(store: &GitStore, commit_hash: &str) -> usize {
    let repo = git2::Repository::open_bare(store.path()).unwrap();
    let oid = git2::Oid::from_str(commit_hash).unwrap();
    let commit = repo.find_commit(oid).unwrap();
    commit.parent_count()
}

/// Helper: return parent commit hashes for a given commit.
fn parent_hashes(store: &GitStore, commit_hash: &str) -> Vec<String> {
    let repo = git2::Repository::open_bare(store.path()).unwrap();
    let oid = git2::Oid::from_str(commit_hash).unwrap();
    let commit = repo.find_commit(oid).unwrap();
    (0..commit.parent_count())
        .map(|i| commit.parent_id(i).unwrap().to_string())
        .collect()
}

// ---------------------------------------------------------------------------
// write with parents
// ---------------------------------------------------------------------------

#[test]
fn write_with_parents_adds_extra_parent() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs_main = store.branches().get("main").unwrap();

    // Create a second branch by pointing it at main's tip, then writing to it.
    store.branches().set("other", &fs_main).unwrap();
    let fs_other = store.branches().get("other").unwrap();
    let fs_other = fs_other
        .write("other.txt", b"other", Default::default())
        .unwrap();

    // Write on main with other's tip as extra parent.
    let new_fs = fs_main
        .write(
            "hello.txt",
            b"hello",
            fs::WriteOptions {
                parents: vec![fs_other.clone()],
                ..Default::default()
            },
        )
        .unwrap();

    let hash = new_fs.commit_hash().unwrap();
    assert_eq!(parent_count(&store, &hash), 2);

    // First parent is the old main tip, second is the other branch tip.
    let parents = parent_hashes(&store, &hash);
    assert_eq!(parents[0], fs_main.commit_hash().unwrap());
    assert_eq!(parents[1], fs_other.commit_hash().unwrap());
}

#[test]
fn write_without_parents_has_single_parent() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();

    let new_fs = fs
        .write("hello.txt", b"hello", Default::default())
        .unwrap();

    let hash = new_fs.commit_hash().unwrap();
    assert_eq!(parent_count(&store, &hash), 1);
}

// ---------------------------------------------------------------------------
// batch with parents
// ---------------------------------------------------------------------------

#[test]
fn batch_with_parents_adds_extra_parent() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs_main = store.branches().get("main").unwrap();

    // Create a second branch.
    store.branches().set("other", &fs_main).unwrap();
    let fs_other = store.branches().get("other").unwrap();
    let fs_other = fs_other
        .write("other.txt", b"other", Default::default())
        .unwrap();

    let mut batch = fs_main.batch(fs::BatchOptions {
        parents: vec![fs_other.clone()],
        ..Default::default()
    });
    batch.write("a.txt", b"aaa").unwrap();
    batch.write("b.txt", b"bbb").unwrap();
    let result = batch.commit().unwrap();

    let hash = result.commit_hash().unwrap();
    assert_eq!(parent_count(&store, &hash), 2);

    let parents = parent_hashes(&store, &hash);
    assert_eq!(parents[1], fs_other.commit_hash().unwrap());
}

// ---------------------------------------------------------------------------
// apply with parents
// ---------------------------------------------------------------------------

#[test]
fn apply_with_parents_adds_extra_parent() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs_main = store.branches().get("main").unwrap();

    store.branches().set("other", &fs_main).unwrap();
    let fs_other = store.branches().get("other").unwrap();
    let fs_other = fs_other
        .write("other.txt", b"other", Default::default())
        .unwrap();

    let new_fs = fs_main
        .apply(
            &[("x.txt", WriteEntry::from_text("x"))],
            &[],
            fs::ApplyOptions {
                parents: vec![fs_other.clone()],
                ..Default::default()
            },
        )
        .unwrap();

    let hash = new_fs.commit_hash().unwrap();
    assert_eq!(parent_count(&store, &hash), 2);
    assert_eq!(parent_hashes(&store, &hash)[1], fs_other.commit_hash().unwrap());
}

// ---------------------------------------------------------------------------
// first-parent lineage preserved
// ---------------------------------------------------------------------------

#[test]
fn first_parent_lineage_preserved_with_extra_parents() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs_main = store.branches().get("main").unwrap();
    let init_hash = fs_main.commit_hash().unwrap();

    store.branches().set("other", &fs_main).unwrap();
    let fs_other = store.branches().get("other").unwrap();
    let fs_other = fs_other
        .write("other.txt", b"other", Default::default())
        .unwrap();

    let new_fs = fs_main
        .write(
            "hello.txt",
            b"hello",
            fs::WriteOptions {
                parents: vec![fs_other],
                ..Default::default()
            },
        )
        .unwrap();

    // parent() walks first-parent lineage and should return the init commit.
    let parent_fs = new_fs.parent().unwrap().unwrap();
    assert_eq!(parent_fs.commit_hash().unwrap(), init_hash);
}

// ---------------------------------------------------------------------------
// multiple extra parents
// ---------------------------------------------------------------------------

#[test]
fn write_with_multiple_extra_parents() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs_main = store.branches().get("main").unwrap();

    store.branches().set("br2", &fs_main).unwrap();
    let fs2 = store.branches().get("br2").unwrap();
    let fs2 = fs2
        .write("f2.txt", b"two", Default::default())
        .unwrap();

    store.branches().set("br3", &fs_main).unwrap();
    let fs3 = store.branches().get("br3").unwrap();
    let fs3 = fs3
        .write("f3.txt", b"three", Default::default())
        .unwrap();

    let new_fs = fs_main
        .write(
            "hello.txt",
            b"hello",
            fs::WriteOptions {
                parents: vec![fs2.clone(), fs3.clone()],
                ..Default::default()
            },
        )
        .unwrap();

    let hash = new_fs.commit_hash().unwrap();
    assert_eq!(parent_count(&store, &hash), 3);

    let parents = parent_hashes(&store, &hash);
    assert_eq!(parents[0], fs_main.commit_hash().unwrap());
    assert_eq!(parents[1], fs2.commit_hash().unwrap());
    assert_eq!(parents[2], fs3.commit_hash().unwrap());
}

// ---------------------------------------------------------------------------
// remove with parents
// ---------------------------------------------------------------------------

// ---------------------------------------------------------------------------
// squash
// ---------------------------------------------------------------------------

#[test]
fn squash_creates_root_commit() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let fs = fs
        .write("hello.txt", b"hello", Default::default())
        .unwrap();

    let squashed = fs.squash(None, None).unwrap();

    // Root commit has zero parents.
    let hash = squashed.commit_hash().unwrap();
    assert_eq!(parent_count(&store, &hash), 0);

    // Tree is preserved.
    assert_eq!(squashed.tree_hash(), fs.tree_hash());

    // Content is readable.
    assert_eq!(squashed.read("hello.txt").unwrap(), b"hello");

    // Detached / read-only.
    assert!(!squashed.writable());
    assert!(squashed.ref_name().is_none());
}

#[test]
fn squash_preserves_tree_hash() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let fs = fs
        .write("a.txt", b"aaa", Default::default())
        .unwrap();
    let fs = fs
        .write("b.txt", b"bbb", Default::default())
        .unwrap();

    let squashed = fs.squash(None, None).unwrap();
    assert_eq!(squashed.tree_hash(), fs.tree_hash());

    // Different commit hash (new commit object).
    assert_ne!(squashed.commit_hash(), fs.commit_hash());
}

#[test]
fn squash_with_parent() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let fs = fs
        .write("hello.txt", b"hello", Default::default())
        .unwrap();

    // Create a parent commit via squash(None).
    let parent = fs.squash(None, None).unwrap();
    // Now squash again with that as parent.
    let child = fs.squash(Some(&parent), None).unwrap();

    let hash = child.commit_hash().unwrap();
    assert_eq!(parent_count(&store, &hash), 1);

    let parents = parent_hashes(&store, &hash);
    assert_eq!(parents[0], parent.commit_hash().unwrap());
}

#[test]
fn squash_with_custom_message() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let fs = fs
        .write("hello.txt", b"hello", Default::default())
        .unwrap();

    let squashed = fs.squash(None, Some("custom squash message")).unwrap();
    let hash = squashed.commit_hash().unwrap();

    // Verify the message via git2.
    let repo = git2::Repository::open_bare(store.path()).unwrap();
    let oid = git2::Oid::from_str(&hash).unwrap();
    let commit = repo.find_commit(oid).unwrap();
    assert_eq!(commit.message().unwrap(), "custom squash message");
}

#[test]
fn squash_assign_to_branch() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let fs = fs
        .write("hello.txt", b"hello", Default::default())
        .unwrap();

    let squashed = fs.squash(None, None).unwrap();

    // Assign to a new branch.
    store.branches().set("squashed", &squashed).unwrap();
    let branch_fs = store.branches().get("squashed").unwrap();
    assert_eq!(branch_fs.tree_hash(), fs.tree_hash());
    assert_eq!(branch_fs.read("hello.txt").unwrap(), b"hello");
    assert_eq!(parent_count(&store, &branch_fs.commit_hash().unwrap()), 0);
}

// ---------------------------------------------------------------------------
// remove with parents
// ---------------------------------------------------------------------------

#[test]
fn remove_with_parents_adds_extra_parent() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs_main = store.branches().get("main").unwrap();

    // Write a file so we have something to remove.
    let fs_main = fs_main
        .write("to_delete.txt", b"gone", Default::default())
        .unwrap();

    // Create a second branch with its own commit.
    store.branches().set("other", &fs_main).unwrap();
    let fs_other = store.branches().get("other").unwrap();
    let fs_other = fs_other
        .write("other.txt", b"other", Default::default())
        .unwrap();

    // Remove with other's tip as extra parent.
    let new_fs = fs_main
        .remove(
            &["to_delete.txt"],
            fs::RemoveOptions {
                parents: vec![fs_other.clone()],
                ..Default::default()
            },
        )
        .unwrap();

    // File should be gone.
    assert!(new_fs.read("to_delete.txt").is_err());

    // Commit should have two parents.
    let hash = new_fs.commit_hash().unwrap();
    assert_eq!(parent_count(&store, &hash), 2);

    let parents = parent_hashes(&store, &hash);
    assert_eq!(parents[0], fs_main.commit_hash().unwrap());
    assert_eq!(parents[1], fs_other.commit_hash().unwrap());
}
